src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.Actions.cs

```diff
@@ -124,11 +127,76 @@ public ProjectAssemblyNameAction(string assemblyName)
                     _assemblyName = assemblyName;
                 }
 
-                public override Task<Compilation> InvokeAsync(Compilation oldCompilation, CancellationToken cancellationToken)
+                public override Task<Compilation> TransformCompilationAsync(Compilation oldCompilation, CancellationToken cancellationToken)
                 {
                     return Task.FromResult(oldCompilation.WithAssemblyName(_assemblyName));
                 }
             }
+
+            internal sealed class AddAnalyzerReferencesAction : CompilationTranslationAction
+            {
+                private readonly IEnumerable<AnalyzerReference> _analyzerReferences;
+
+                public AddAnalyzerReferencesAction(IEnumerable<AnalyzerReference> analyzerReferences)
+                {
+                    _analyzerReferences = analyzerReferences;
+                }
+
+                public override TrackedGeneratorDriver TransformGeneratorDriver(TrackedGeneratorDriver generatorDriver)
+                {
+                    var generators = _analyzerReferences.SelectMany(a => a.GetGenerators()).ToImmutableArray();
+                    return new TrackedGeneratorDriver(generatorDriver.GeneratorDriver?.AddGenerators(generators));
+                }
+            }
+
+            internal sealed class RemoveAnalyzerReferencesAction : CompilationTranslationAction
+            {
+                private readonly ImmutableArray<AnalyzerReference> _analyzerReferences;
+
+                public RemoveAnalyzerReferencesAction(ImmutableArray<AnalyzerReference> analyzerReferences)
+                {
+                    _analyzerReferences = analyzerReferences;
+                }
+
+                public override TrackedGeneratorDriver TransformGeneratorDriver(TrackedGeneratorDriver generatorDriver)
+                {
+                    var generators = _analyzerReferences.SelectMany(a => a.GetGenerators()).ToImmutableArray();
+                    return new TrackedGeneratorDriver(generatorDriver.GeneratorDriver?.RemoveGenerators(generators));
+                }
+            }
+
+            internal sealed class AddAdditionalDocumentsAction : CompilationTranslationAction
```

Comment:
base name is a bit off now.  CompilationTranslationAction.. but it translates compilations and generators.  

do we always have to execute both translation steps in lockstep?  if so, might be better ot have this return a pair of the new compilation and new generator driver.

Reply:
We don't _have_ to do them both in lockstep but both would be holding onto the same data, and potentially running one but not the other means we need to deal with issues where one has been pulled forward but the other hasn't.

I'll rename the base type in any case.

Reply:
Renamed.